src/irmin-layers/stats.ml

```diff
@@ -15,113 +15,187 @@
-}
+(* This file avoids [option] type and clean functionnal paradigms in order to
+   lower the cpu footprint *)
+
+(** Ensure lists are not growing indefinitely by dropping elements. *)
+let limit_length_list = 10
 
-type i = {
+type freeze_profile = {
+  mutable idx : int;
+  mutable t0 : Mtime.t;
+  mutable t1 : Mtime.t;
+  mutable current_section : string;
+  mutable rev_timeline : (string * Mtime.t * float) list;
   mutable contents : int;
   mutable nodes : int;
   mutable commits : int;
   mutable branches : int;
   mutable adds : int;
+  mutable skips : int;
+  mutable copy_newies_loops : int;
+  mutable yields : int;
+  mutable outside_totlen : float;
+  mutable inside_totlen : float;
+  mutable outside_maxlen : string * float;
+  mutable inside_maxlen : string * float;
 }
 
-let fresh_stats_t () =
+let fresh_freeze_profile idx t0 current_section =
   {
-    nb_freeze = 0;
-    copied_contents = [];
-    copied_nodes = [];
-    copied_commits = [];
-    copied_branches = [];
-    waiting_freeze = [];
-    completed_freeze = [];
+    idx;
+    t0;
+    t1 = t0;
+    current_section;
+    rev_timeline = [];
+    contents = 0;
+    nodes = 0;
+    commits = 0;
+    branches = 0;
+    adds = 0;
     skips = 0;
+    copy_newies_loops = 0;
+    yields = 0;
+    outside_totlen = 0.;
+    inside_totlen = 0.;
+    outside_maxlen = ("never", 0.);
+    inside_maxlen = ("never", 0.);
   }
 
-let fresh_stats_i () =
-  { contents = 0; nodes = 0; commits = 0; branches = 0; adds = 0 }
+let get_elapsed =
+  let c = ref (Mtime_clock.counter ()) in
+  fun ~reset ->
+    let elapsed = Mtime.Span.to_s (Mtime_clock.count !c) in
+    if reset then c := Mtime_clock.counter ();
+    elapsed
 
-let stats_t = fresh_stats_t ()
-let stats_i = fresh_stats_i ()
+let freeze_start_counter =
+  let c = ref 0 in
+  fun () ->
+    incr c;
+    !c - 1
 
-let reset_stats_i () =
-  stats_i.contents <- 0;
-  stats_i.nodes <- 0;
-  stats_i.commits <- 0;
-  stats_i.branches <- 0
+let freeze_profiles = ref []
+let latest = ref (fresh_freeze_profile ~-1 (Mtime_clock.now ()) "")
 
 let reset_stats () =
-  stats_t.nb_freeze <- 0;
-  stats_t.copied_contents <- [];
-  stats_t.copied_nodes <- [];
-  stats_t.copied_commits <- [];
-  stats_t.copied_branches <- [];
-  stats_t.skips <- 0;
-  reset_stats_i ()
-
-let get () =
-  {
-    nb_freeze = stats_t.nb_freeze;
-    copied_contents = stats_i.contents :: stats_t.copied_contents;
-    copied_nodes = stats_i.nodes :: stats_t.copied_nodes;
-    copied_commits = stats_i.commits :: stats_t.copied_commits;
-    copied_branches = stats_i.branches :: stats_t.copied_branches;
-    waiting_freeze = stats_t.waiting_freeze;
-    completed_freeze = stats_t.completed_freeze;
-    skips = stats_t.skips;
-  }
+  freeze_profiles := [];
+  latest := fresh_freeze_profile ~-1 (Mtime_clock.now ()) ""
 
-(** Ensure lists are not growing indefinitely by dropping elements. *)
-let limit_length_list = 10
+let freeze_start t0 current_section =
```

Comment:
you can replace `current_section` here with "wait for freeze lock"

Reply:
I renamed the parameter to `initial_section`. I like having all the section names hard-coded in a single file, i.e. ext_layered.ml